src/gui/painting/qpainter.cpp

```diff
@@ -2673,6 +2673,183 @@ QPainterPath QPainter::clipPath() const
+        path = QPainterPath();
+    } else {
+        QPainterPath tempPath;
+        tempPath.addRect(info.rectf);
+
+        path = tempPath * matrix;
+    }
+}
+
+static void join_path(QPainterPath &path, QTransform matrix, bool &lastWasNothing, const QPainterClipInfo info) {
+    if (lastWasNothing) {
+        path = info.path * matrix;
+        lastWasNothing = false;
+        return;
+    }
+    if (info.operation == Qt::IntersectClip) {
+        path &= (info.path * matrix);
+    } else if (info.operation == Qt::NoClip) {
+        lastWasNothing = true;
+        path = QPainterPath();
+    } else {
+        path = info.path * matrix;
+    }
+}
+
+static void join_rect(QPainterPath &path, QTransform matrix, bool &lastWasNothing, const QPainterClipInfo info) {
+    if (lastWasNothing) {
+        QPainterPath tempPath;
+        tempPath.addRect(info.rect);
+        tempPath = tempPath * matrix;
+
+        path = tempPath;
+        lastWasNothing = false;
+        return;
+    }
+    if (info.operation == Qt::IntersectClip) {
+        QPainterPath tempPath;
+        tempPath.addRect(info.rect);
+        tempPath = tempPath * matrix;
+
+        path &= tempPath;
+    } else if (info.operation == Qt::NoClip) {
+        lastWasNothing = true;
+        path = QPainterPath();
+    } else {
+        QPainterPath tempPath;
+        tempPath.addRect(info.rect);
+        tempPath = tempPath * matrix;
+
+        path = tempPath;
+    }
+}
+
+/*!
+    Returns the current clip path in logical coordinates, unrounded.
+
+    The main purpose for this function is inside QSvgGenerator,
+    where we do not want any points to be rounded during the
+    creation of the file.
+
+    \warning QPainter does not store the combined clip explicitly as
+    this is handled by the underlying QPaintEngine, so the path is
+    recreated on demandand transformed to the current logical
+    coordinate system.This is potentially an expensive operation.
+
+    \sa setClipPath(), clipRegion(), setClipping()
+*/
+QPainterPath QPainter::clipPathF() const
+{
+    Q_D(const QPainter);
+
+    if (!d->engine) {
+        qWarning("QPainter::clipPath: Painter not active");
+        return QPainterPath();
+    }
+
+    // No clip, return empty
+    if (d->state->clipInfo.isEmpty()) {
+        return QPainterPath();
+    } else {
+
+        // Update inverse matrix, used below.
+        if (!d->txinv)
+            const_cast<QPainter *>(this)->d_ptr->updateInvMatrix();
+
+        /*
+        * Modified version of "clipRegion" that is used in ::clipPath
+        * as "return qt_regionToPath(clipRegion());" that prevents rounding of values,
+        * placing them directly into a QPainterPath.
+        */
+        QPainterPath path;
+        bool lastWasNothing = true;
+        if (!d->txinv)
+            const_cast<QPainter *>(this)->d_ptr->updateInvMatrix();
+
+        // ### Falcon: Use QPainterPath
+        for (const QPainterClipInfo &info : qAsConst(d->state->clipInfo)) {
```

Comment:
After you restructured your code it became quite clear that regardless of of clip type actions that you are taking follow common pattern. I think it would be better if you keep the function flow here and instead create several small functions that will perform action specific to clip path.

You can define this action functions this way:
```
static QPainterPath noClipping(const QTransform & matrix, const QPainterClipInfo & info)
{
	return QPainterPath();
}

static QPainterPath PathClipInitial(const QTransform & matrix, const QPainterClipInfo & info)
{
	return PathClipReplaceClip(matrix, info);
}

static QPainterPath PathClipReplaceClip(const QTransform & matrix, const QPainterClipInfo & info)
{
	return info.path * matrix;
}

static QPainterPath PathClipIntersectClip(const QTransform & matrix, const QPainterClipInfo & info)
{
	return PathClipReplaceClip(matrix, info);
}

static QPainterPath RectClipInitial(const QTransform & matrix, const QPainterClipInfo & info)
{
	return RectClipReplaceClip(matrix, info);
}

static QPainterPath RectClipReplaceClip(const QTransform & matrix, const QPainterClipInfo & info)
{
	QPainterPath tempPath;
    tempPath.addRect(info.rect);
    tempPath = tempPath * matrix;
	return tempPath;
}

static QPainterPath RectClipIntersectClip(const QTransform & matrix, const QPainterClipInfo & info)
{
	return RectClipReplaceClip(matrix, info);
}

static QPainterPath RectFClipInitial(const QTransform & matrix, const QPainterClipInfo & info)
{
	return RectFClipReplaceClip(matrix, info);
}

static QPainterPath RectFClipReplaceClip(const QTransform & matrix, const QPainterClipInfo & info)
{
	QPainterPath tempPath;
    tempPath.addRect(info.rectf);
	return tempPath;
}

static QPainterPath RectFClipIntersectClip(const QTransform & matrix, const QPainterClipInfo & info)
{
	//if you simplify this method, you can simplify clipped_path
	// Use rect intersection if possible.
    if (matrix.type() <= QTransform::TxScale) 
	{
		QPainterPath tempPath;
        tempPath.addRect(matrix.mapRect(info.rectf));
    } 
	else 
	{
        QPainterPath tempPath;
        tempPath.addRegion(matrix.map(QRegion(info.rectf.toRect())));
    }
	return tempPath;
}

static QPainterPath RegionClipInitial(const QTransform & matrix, const QPainterClipInfo & info)
{
	return RegionClipReplaceClip(matrix, info);
}

static QPainterPath RegionClipReplaceClip(const QTransform & matrix, const QPainterClipInfo & info)
{
    QPainterPath tempPath;
    tempPath.addRegion(info.region * matrix);
	return tempPath;
}

static QPainterPath RegionClipIntersectClip(const QTransform & matrix, const QPainterClipInfo & info)
{
	return RegionClipReplaceClip(matrix, info);
}

```

Now you need to defined mapping for actions to be taken:
```
enum ActionForClipOperation {LastWasNothing, NoClip, ReplaceClip, IntersectClip}
typedef QPainterPath (*PainterFunc)(const QTransform & matrix, const QPainterClipInfo & info);
PainterFunc clipped_path[][] = {
{RegionClipInitial, noClipping, RegionClipReplaceClip, RegionClipIntersectClip},
{PathClipInitial, noClipping, PathClipReplaceClip, PathClipIntersectClip},
{RectClipInitial, noClipping, RectClipReplaceClip, RectClipIntersectClip},
{RectFClipInitial, noClipping, RectFClipReplaceClip, RectFClipIntersectClip}
};
```

After that your for look should look like this:

```
for (const QPainterClipInfo &info : qAsConst(d->state->clipInfo)) 
{
    QTransform matrix = (info.matrix * d->invMatrix);
	
    if (lastWasNothing) {
        path = clipped_path[info.clipType][LastWasNothing];
        lastWasNothing = false;
        continue;
    }
	switch (info.operation)
	{
		case Qt::IntersectClip:
		{
			path &= clipped_path[info.clipType][IntersectClip];
			break;
		}
		case Qt::NoClip:
		{
			lastWasNothing = true;
			path = clipped_path[info.clipType][NoClip];
			break;
		}
		case Qt::ReplaceClip:
		{
			path = clipped_path[info.clipType][ReplaceClip];
			break;
		}
	}
}

```
Please note, that I haven't complied this code, it is just an idea/direction, but I think it should be pretty easy to fix compiler errors if they arise. Let me know if you have any questions. Also feel free to change the naming I used if you find better names.